python/pyiceberg/expressions/base.py

```diff
@@ -342,151 +313,197 @@ def __str__(self) -> str:
 
 
 @dataclass(frozen=True)
-class AlwaysTrue(BooleanExpression, ABC, Singleton):
+class AlwaysTrue(BooleanExpression, Singleton):
     """TRUE expression"""
 
     def __invert__(self) -> AlwaysFalse:
         return AlwaysFalse()
 
 
 @dataclass(frozen=True)
-class AlwaysFalse(BooleanExpression, ABC, Singleton):
+class AlwaysFalse(BooleanExpression, Singleton):
     """FALSE expression"""
 
     def __invert__(self) -> AlwaysTrue:
         return AlwaysTrue()
 
 
-class IsNull(UnboundPredicate[T]):
-    def __invert__(self) -> NotNull:
-        return NotNull(self.term)
+@dataclass(frozen=True)
```

Comment:
An abstract class should not be a dataclass as well, since it cannot be instantiated, it does not carry any data.

Reply:
If we make `__invert__` abstract, then we can't use a frozen `@dataclass` without duplicating a lot more.

The problem is, as you pointed out, `@dataclass` can't be abstract. If I move this to the child class, `IsNull`, then the constructor no longer supports `term` because it is defined here in the parent. Then we have to implement `__invert__` in each child class and define the `term` separately in each child class.

That basically removes the benefit of using a common base class. We can no longer rely on anything higher than the leaf class having a `term` or a set of `literals`.

I think I'd rather trade the mypy hints for `__invert__` than not be able to rely on inheritance for expressions. That is, if I have a `UnaryPredicate`, I want to be able to use its `term` instead of handling everything at the leaf level.

Reply:
I solved this by implementing `__invert__` and raising `NotImplementedError`. It's overridden in subclasses just like before, but now `@dataclass` works because these aren't abstract.